swap/simulations_test.go

```diff
@@ -0,0 +1,779 @@
+		}
+
+		return nil
+
+	})
+
+	if result.Error != nil {
+		t.Fatal(result.Error)
+	}
+
+	log.Info("Simulation ended")
+}
+
+// TestBasicSwapSimulation starts 16 nodes, then in a simple round robin fashion sends messages to each other.
+// Then checks that accounting is ok. It checks the actual amount of balances without any cheques sent,
+// in order to verify that the most basic accounting works.
+func TestBasicSwapSimulation(t *testing.T) {
+	nodeCount := 16
+	// create the shared backend and params
+	params, err := newSharedBackendSwaps(nodeCount)
+	if err != nil {
+		t.Fatal(err)
+	}
+	// cleanup backend
+	defer params.backend.Close()
+
+	// initialize the simulation
+	sim := simulation.NewInProc(newSimServiceMap(params))
+	defer sim.Close()
+
+	log.Info("Initializing")
+
+	ctx, cancelSimRun := context.WithTimeout(context.Background(), 1*time.Minute)
+	defer cancelSimRun()
+
+	_, err = sim.AddNodesAndConnectFull(nodeCount)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	log.Debug("Wait for all connections to be established")
+	// first check: let's make sure that all nodes have been connected to the others
+	// this is the maximum number of possible connections
+	maxConns := nodeCount * (nodeCount - 1) / 2
+CONNS:
+	for {
+		select {
+		// let's be nice and make sure we catch a timeout
+		case <-ctx.Done():
+			t.Fatal("Timed out waiting for all connections to be established")
+		default:
+			// this should be true when all connections have been established
+			if len(sim.Net.Conns) == maxConns {
+				break CONNS
+			}
+		}
+		// don't overheat the CPU...
+		time.Sleep(5 * time.Millisecond)
+	}
+
+	log.Info("starting simulation...")
+
+	// we don't want any cheques to be issued for this test, we only want to test accounting across nodes
+	// for this we define a "global" maximum amount of messages to be sent;
+	// this formula should ensure that we trigger enough messages but not enough to trigger cheques
+	maxMsgs := (DefaultPaymentThreshold / params.maxMsgPrice) * (nodeCount - 1)
+
+	// need some synchronization to make sure we wait enough before checking all balances:
+	// all messages should have been received, otherwise there may be some imbalances!
+	allMessagesArrived := make(chan struct{})
+
+	metricsReg := metrics.AccountingRegistry
+	cter := metricsReg.Get("account.msg.credit")
+	counter := cter.(metrics.Counter)
+	counter.Clear()
+
+	go func() {
+		for {
+			maxMsgsInt64 := int64(maxMsgs)
+			select {
+			case <-ctx.Done():
+				return
+			default:
+			}
+			// all messages have been received
+			if counter.Count() == maxMsgsInt64 {
+				close(allMessagesArrived)
+				return
+			}
+			time.Sleep(10 * time.Millisecond)
+		}
+	}()
+
+	result := sim.Run(ctx, func(ctx context.Context, sim *simulation.Simulation) (err error) {
+		log.Info("simulation running")
+
+		ctx, cancel := context.WithTimeout(context.Background(), 60*time.Second)
+		defer cancel()
+		_, err = sim.WaitTillHealthy(ctx)
```

Comment:
Since further up we already verify for full connectivity this check might not be necessary.